beancount.el

```diff
@@ -876,14 +878,42 @@ what that column is and returns it (an integer)."
   "Start a new timestamped directive with date DAYS before today."
   (interactive "P")
   (unless (bolp) (newline))
-  (insert (beancount--shift-current-date days) " "))
-
-(defun beancount--shift-current-date (days)
-  "Return ISO-8601 formatted date DAYS before today."
-  (let ((days-to-shift (- (or days 0))))
-    (format-time-string
-     "%Y-%m-%d"
-     (time-add (current-time) (days-to-time days-to-shift)))))
+  (insert (beancount--shift-date (current-time) (- (or days 0))) " "))
+
+(defun beancount--shift-date (today days)
+  "Return ISO-8601 formatted date TODAY shifted by DAYS.
+DAYS can negative or positive."
+  (format-time-string
+   "%Y-%m-%d"
+   (time-add today (days-to-time days))))
+
+(defun beancount--shift-date-at-point (days)
+  "Shift the date under point by a specified number of DAYS."
+  (let* ((date-regexp beancount-date-regexp)
+         (current-line (thing-at-point 'line t))
+         (pos (- (point) (pos-bol)))
+         (date (if (and (string-match date-regexp current-line)
+                        (>= pos (match-beginning 0))
+                        (<= pos (match-end 0)))
+                   (match-string 0 current-line)
+                 (user-error "No date found under point")))
+         (newdate (beancount--shift-date (date-to-time date) days)))
+    (save-excursion
+      (beginning-of-line)
+      (when (re-search-forward date-regexp (line-end-position) t)
+        (replace-match newdate t t nil)))))
```

Comment:
I think this can be simplified to
```elisp
(defun beancount--shift-date-at-point (days)
  "Shift the date under point by a specified number of DAYS."
  (save-excursion
    (if (thing-at-point-looking-at beancount-date-regexp 10)
      (let ((newdate (beancount--shift-date (save-match-data (date-to-time (match-string-no-properties 0))) days)))
        (replace-match newdate t t))
      (user-error "No date at point"))))
```
This gets simpler with the changes to `beancount--shift-date` proposed above, so much that the `let` may not be helping readability that much anymore.

With this simplification, I'm not sure that `(save-excursion ...)` is needed.